### PR TITLE
rustc_llvm: adapt to flattened CLI args in LLVM

### PR DESCRIPTION
**Pull request by durin42**
This changed in
llvm/llvm-project@e190d074a0a77c9f8a7d7938a8187a7e2076e290. I decided to stick with more duplication between the ifdef blocks to make the code easier to read for the next two years before we can plausibly drop LLVM 19.

@rustbot label: +llvm-main

try-job: x86_64-msvc

### CODE REVIEW

**Comment by nebulark** [comment]
Was already going to make a PR for this, as this is due to my change in LLVM, but you were faster.
I think it might make sense to expose the Argv0 and the flattened commandline as args for LLVMRustCreateTargetMachine. That way we can format the string on the rust. What do you think?

Still wip but something like this: https://github.com/rust-lang/rust/compare/master...nebulark:rust_contributing:fix_cl

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp [hunk 1]
```diff
@@ -498,6 +498,21 @@ extern "C" LLVMTargetMachineRef LLVMRustCreateTargetMachine(
   Options.EmitStackSizeSection = EmitStackSizeSection;
 
   if (ArgsCstrBuff != nullptr) {
+#if LLVM_VERSION_GE(20, 0)
+    int buffer_offset = 0;
+    assert(ArgsCstrBuff[ArgsCstrBuffLen - 1] == '\0');
+    std::string Arg0(ArgsCstrBuff);
+    buffer_offset = Arg0.size() + 1;
+    std::string ArgsCppStr(ArgsCstrBuff + buffer_offset, ArgsCstrBuffLen - 1);
```
Comment (workingjubilee) [inline compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:506]: So, this PR looks overall fine but this style of constructor is recommended against by several C++ style guides, including e.g. [Abseil's](https://abseil.io/tips/88), which is why I removed all instances of it from rustc_llvm in favor of assignment constructors, which look like... well, more like all the rest of our code in this repo: https://github.com/rust-lang/rust/pull/122062

Comment (durin42) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp]: Fair enough - I never can remember the C++ right choices. Fixed.

**Comment by durin42** [comment]
> Was already going to make a PR for this, as this is due to my change in LLVM, but you were faster. I think it might make sense to expose the Argv0 and the flattened commandline as args for LLVMRustCreateTargetMachine. That way we can format the string on the rust. What do you think?
> 
> Still wip but something like this: [master...nebulark:rust_contributing:fix_cl](https://github.com/rust-lang/rust/compare/master...nebulark:rust_contributing:fix_cl)

I don't really feel strongly - I was mostly trying to not change the interface between the Rust and C++ code until we could really clean house on the old codepath, at which point I'd think harder about it.

We've got a [continuous build](https://buildkite.com/llvm-project/rust-llvm-integrate-prototype/builds/) that builds Rust HEAD against LLVM HEAD as that's what we use internally, so we noticed pretty quickly.

**Review by workingjubilee — APPROVED** [review]

**Comment by workingjubilee** [comment]
heh, clang-format

**Comment by nebulark** [comment]
Had another look. There is now a difference in output (besides the now missing erronous "cc1" part). I'd expect this windows-only test to fail:
https://github.com/rust-lang/rust/blob/master/tests/run-make/pdb-buildinfo-cl-cmd/rmake.rs

Previously the arguments were quoted via  `llvm::sys::printArg`: 
https://github.com/llvm/llvm-project/blob/41f1b467a29d2ca4e35df37c3aa79a0a8c04bc4f/clang/lib/CodeGen/BackendUtil.cpp#L325-L357
https://github.com/llvm/llvm-project/blob/d703b922961e0d02a5effdd4bfbb23ad50a3cc9f/llvm/lib/Support/Program.cpp#L83-L99

For me it's fine though to leave it as it is, so everything compiles again. I'll make a follow up pr in this case.

**Comment by workingjubilee** [comment]
hm!

@bors r+

**Pull request merged**